File: elixir/lib/days/day06.ex
defmodule Days.Day06 do
  @moduledoc """
  Solutions to AoC 2022 Day 6: Tuning Trouble.
  """

  # Public Functions
  @spec pt1(binary()) :: integer()
  @spec pt2(binary()) :: integer()

  # Private Functions
  @spec find_marker(list(), integer()) :: integer()

  @doc """
  Part 1: Find start-of-packet marker.

  ## Examples

    iex> Days.Day06.pt1(day06_input) # filename
    <Solution to Day 6, part 1>

  """
  def pt1(file_name) do
    File.read!(file_name)
      |> String.to_charlist()
      |> find_marker(4)
  end

  @doc """
  Part 2: Find start-of-messaged marker.

  ## Examples

    iex> Days.Day06.pt2(day06_input) # filename
    <Solution to Day 6, part 2>

  """
  def pt2(file_name) do
    File.read!(file_name)
      |> String.to_charlist()
      |> find_marker(14)
  end

  # --------------- Private Functions --------------- #

  defp find_marker(list, marker_size) do
    list
      |> Enum.chunk_every(marker_size, 1, :discard)
      |> Enum.with_index()
      |> Enum.find(fn {clist, _i} -> Enum.uniq(clist) == clist end)
      |> elem(1)
      |> (&(&1 + marker_size)).()
  end
end

File: python/days/day06.py
"""Day 6 of AOC 2022: Tuning Trouble"""

# -------------- Public Functions ------------- #

def day06_pt1(filename: str) -> int:
    """Day 6, part 1: Find start-of-packet marker"""
    with open(filename, 'r', encoding='utf-8') as input_file:
        stream = input_file.read()

    return _find_marker(stream, 4)

def day06_pt2(filename: str) -> int:
    """Day 6, part 2: Find start-of-message marker"""
    with open(filename, 'r', encoding='utf-8') as input_file:
        stream = input_file.read()

    return _find_marker(stream, 14)


# ----------------- Constants ----------------- #


# ------------- Private Functions ------------- #

def _is_str_uniq(string: str) -> bool:
    chars = set()
    for char in string:
        if char in chars:
            return False
        else:
            chars.add(char)

    return True

def _find_marker(stream: str, marker_size: int, identifier=_is_str_uniq) -> int:
    sliding_window = (stream[i:i+marker_size] for i in range(len(stream)-marker_size))
    for i, marker in  enumerate(sliding_window):
        if identifier(marker):
            return i + marker_size



# ------------------- Script ------------------ #

if __name__ == '__main__':
    INPUT_FILENAME = '../inputs/06.txt'
    print(f'Day 6, pt1: {day06_pt1(INPUT_FILENAME)}')
    print(f'Day 6, pt2: {day06_pt2(INPUT_FILENAME)}')
